### PR TITLE
adding plumed support + generic non-gmxapi version

### PR DESCRIPTION
**Pull request by milafternoon**
Hi, please take your time to revise the changes & feel free to suggest changes. 

I'll add some extra docs soon to account for the additional options, but for now the existing functionalities should not be affected.

### CODE REVIEW

File: mdtools.py [hunk 1]
```diff
@@ -48,46 +58,57 @@ def mdrun(
     check_point_file: str = None,
     mdrun_options: list = None,
     gpus_per_node: int = None,
+    plumed_file: str = None,
+    use_api: bool = True
 ):
-    mpi_rank = max(mpi_rank - 1, 0)
+    mpi_rank = mpi_rank - 1
```
Comment (sperezconesa) [inline mdtools.py:64]: I added this to make it work with one mpirank.
```suggestion
    mpi_rank = max(mpi_rank - 1,0)
```

File: mdtools.py [hunk 2]
```diff
@@ -48,46 +58,57 @@ def mdrun(
...
         input_files["-cpi"] = check_point_file
+    if plumed_file is not None:
+        input_files["-plumed"] = plumed_file
     # SPC increased state printing to every 5 minutes since swarms are short
-    if mdrun_options is None:
+    if mdrun_options is None:  # TODO
         mdrun_options_parse = []
     else:
-        mdrun_options_parse = mdrun_options[:]
+        mdrun_options_parse = mdrun_options
```
Comment (sperezconesa) [inline mdtools.py:76]: Not adding this generated a bug.
```suggestion
        mdrun_options_parse = mdrun_options[:]
```

**Review by sperezconesa — CHANGES_REQUESTED** [review]

**Review by sperezconesa — APPROVED** [review]

**Comment by sperezconesa** [comment]
Looks good, I'll start testing ASAP and get back to you.

**Comment by milafternoon** [comment]
Sure, here I'm finally running it with my system of interest, I'll
hopefully catch any issues should they come up! Also working on separating
the example into two distinct directories and adding a small section to
README.


wt., 18 maj 2021 o 14:45 Sergio Pérez Conesa ***@***.***>
napisał(a):

**Comment by sperezconesa** [comment]
How did you run the simulations with gmx_mpi and no gmxapi? I'm having trouble doing that. What command did you add to `slurm.sh` and what did you modify in `mdtools.py`?

**Comment by milafternoon** [comment]
Hi, I tried to clarify that in a follow-up mail - I did not in fact manage
to do that, at some point I mixed up two versions and it was gmx that
worked all the time, not gmx_mpi.



El dt., 18 de maig 2021, 18:32, Sergio Pérez Conesa <
***@***.***> va escriure:

**Review by sperezconesa — APPROVED** [review]

**Review by sperezconesa — COMMENTED** [review]
image
This gives me a syntax error. I don't understand why. I'm using python 3.8.5

**Comment by milafternoon** [comment]
Sorry, should be fixed now, I accidentally wrote "line in ..." instead of "line for line in..." while splitting an unreadable one-liner into two list comprehensions.

File: mdtools.py [hunk 3]
```diff
@@ -26,17 +34,36 @@ def grompp(
...
+            arguments=["grompp"] + grompp_options,
+            input_files=input_files,
+            output_files=output_files,
+        )
+        prep.run()
+        output = str(prep.output.erroroutput.result()).strip()
+    else:
+        infiles = ' '.join([k + ' ' + v for k, v in input_files.items()])
+        outfiles = ' '.join([k + ' ' + v for k, v in output_files.items()])
+        gmx = "mpiexec -n 1 gmx_mpi" if gmx_executable == "gmx_mpi" else "gmx"
```
Comment (sperezconesa) [inline mdtools.py:49]: So, mpiexec in my cluster needs to be srun... maybe there should be an input variable for this in config.json

File: mdtools.py [hunk 4]
```diff
@@ -26,17 +34,36 @@ def grompp(
...
+        output = result.stderr
+
     if output:
         logger.info("grompp output:\n%s", output)
-    return prep
+
+
+def grompp_all(task_list: List[dict]):
+    from multiprocessing import Pool
+    proc_per_core = int(os.environ["SLURM_CPUS_ON_NODE"]) if "SLURM_CPUS_ON_NODE" in os.environ.keys() else 1
```
Comment (sperezconesa) [inline mdtools.py:60]: This should be SLURM_NTASKS_PER_NODE which is equal to the number of mpi tasks you can do in the node. The current variable is equal to the number of threads in the node, which in your case is equal to the number of mpitasks.

File: mdtools.py [hunk 5]
```diff
@@ -26,17 +34,36 @@ def grompp(
...
+    with Pool(proc_per_core) as p:
+        p.map(_grompp, task_list)
+
+
+def _grompp(args: dict):
```
Comment (sperezconesa) [inline mdtools.py:65]: args.update({'gmx_executable': 'gmx_mpi'}) in my system returns None and does the update to args. Maybe this has changed from python version (I'm using 3.8.5).
I suggest:
```python
def _grompp(args: dict):
    args.update({"gmx_executable": "gmx_mpi"})
    return grompp(**args)
```

**Review by sperezconesa — COMMENTED** [review]
These changes had to be done for it to work on my cluster.

**Review by sperezconesa — COMMENTED** [review]
Sorry I didn't talk about this with you. I am using black -l 79 for formatting. But I don't have any strong opinions so we can use something else if you want.

**Comment by sperezconesa** [comment]
I have decide to make a dev branch in the main repo so me and Milosz can work more efficiently.

**Pull request merged**